llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java

```diff
@@ -0,0 +1,114 @@
+ * and the attached local disk might end up being the same when a new daemon starts (this applies to
+ * on-prem as well as cloud scenarios).
+ */
+public class LocalDirCleaner extends AbstractService {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalDirCleaner.class);
+
+  private String[] localDirs;
+
+  private long cleanupIntervalSec;
+  private long fileModifyTimeThresholdSec;
+
+  ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+  public LocalDirCleaner(String[] localDirs, Configuration conf) {
+    super("LocalDirCleaner");
+    this.localDirs = localDirs;
+    this.cleanupIntervalSec = getInterval(conf);
+    this.fileModifyTimeThresholdSec = getFileModifyTimeThreshold(conf);
+    LOG.info("Initialized local dir cleaner: interval: {}s, threshold: {}s", cleanupIntervalSec,
+        fileModifyTimeThresholdSec);
+  }
+
+  @Override
+  public void serviceStart() throws IOException {
+    scheduler.scheduleAtFixedRate(this::cleanup, 0, cleanupIntervalSec, TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void serviceStop() throws IOException {
+    // we can shutdown this service now and ignore leftovers, because under normal circumstances,
+    // files from the local dirs are cleaned up (so LocalDirCleaner is a best effort utility)
+    scheduler.shutdownNow();
+  }
+
+  private long getFileModifyTimeThreshold(Configuration conf) {
+    return HiveConf.getTimeVar(conf, HiveConf.ConfVars.LLAP_LOCAL_DIR_CLEANER_FILE_MODIFY_TIME_THRESHOLD,
+        TimeUnit.SECONDS);
+  }
+
+  private long getInterval(Configuration conf) {
+    return HiveConf.getTimeVar(conf, HiveConf.ConfVars.LLAP_LOCAL_DIR_CLEANER_CLEANUP_INTERVAL, TimeUnit.SECONDS);
+  }
+
+  private void cleanup() {
+    Instant deleteBefore = Instant.now().minus(fileModifyTimeThresholdSec, ChronoUnit.SECONDS);
+
+    for (String localDir : localDirs) {
+      Path pathLocalDir = Paths.get(localDir);
+      cleanupPath(deleteBefore, pathLocalDir);
+    }
+  }
+
+  private void cleanupPath(Instant deleteBefore, Path pathLocalDir) {
+    LOG.info("Cleaning up files older than {} from {}", deleteBefore, pathLocalDir);
+
+    try (Stream<Path> files = Files.walk(pathLocalDir)) {
+      files.filter(f -> {
+        try {
+          FileTime modified = Files.getLastModifiedTime(f);
+          LOG.debug("Checking: {}, modified: {}", f, modified);
+          return Files.isRegularFile(f) && modified.toInstant().isBefore(deleteBefore);
+        } catch (IOException ex) {
+          LOG.warn("IOException caught while checking file for deletion", ex);
+          return false;
+        }
+      }).forEach(f -> {
+        try {
+          LOG.info("Delete old local file: {}", f);
+          Files.delete(f);
+        } catch (IOException ex) {
+          LOG.warn("Failed to delete file", ex);
+        }
+      });
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```

Comment:
Maybe return a more user-friendly exception? And aslo this cleaner is not a very necessary thread, maybe we can just `LOG.warn/error` instead of throw exception?

Reply:
right, logging is enough, we're not handling this exception anyway